### PR TITLE
Add Timestamp support to CachedResultSet getTime() and getDate() methods + Complete missing test coverage

### PR DESCRIPTION
**Pull request by Frank-Gu-81**
### Summary

Add Timestamp support to CachedResultSet getTime() and getDate() methods + Complete missing test coverage

### Description

Problem: CachedResultSet was missing proper handling of Timestamp objects in convertToTime() and convertToDate() methods, causing them to fall through to string parsing instead of proper type conversion. Additionally, several getter methods lacked comprehensive unit test coverage.

Changes Made:

Core Functionality:
• **Fixed convertToTime() method**: Added instanceof Timestamp case that uses new Time(timestamp.getTime()) for standard JDBC behavior
• **Fixed convertToDate() method**: Added instanceof Timestamp case that uses new Date(timestamp.getTime()) for standard JDBC behavior  
• **Enhanced test coverage**: Added comprehensive test cases in test_get_special_time() for Timestamp conversion scenarios including Calendar timezone handling
• **Fixed test object equality**: Resolved test failures by using consistent constructors (new Time() vs Time.valueOf())

Test Coverage Improvements (per CachedResultSet-Test-Coverage-Plan.md):
• **✅ Completed getBoolean() tests**: Added comprehensive type conversion tests (numbers, strings, null handling)
• **✅ Completed getBytes() tests**: Added tests for byte arrays, non-byte inputs, null handling
• **✅ Completed getURL() tests**: Added tests for URL objects, valid/invalid URL strings, SQLException handling
• **✅ Completed getNString() tests**: Added tests delegating to getString() with proper null handling
• **✅ Completed getObject(Class<T>) tests**: Added generic type-safe retrieval tests for both index and label versions
• **✅ Completed wrapper interface tests**: Added unwrap() and isWrapperFor() tests for interface compliance

Test Coverage: All tests pass including edge cases for epoch timestamps, timezone conversions, type conversions, and null handling across all getter methods.

### Note: Implementation Choice

Direct millisecond preservation approach (new Time(timestamp.getTime())) was used rather than PostgreSQL JDBC's rebasing behavior, providing more intuitive timezone handling that avoids DST-related 1-hour shifts.

### CODE REVIEW

File: wrapper/src/main/java/software/amazon/jdbc/plugin/cache/CachedResultSet.java [hunk 1]
```diff
@@ -268,6 +269,14 @@ private Date convertToDate(Object dateObj, Calendar cal) throws SQLException {
       ZonedDateTime targetZonedDateTime = originalZonedDateTime.withZoneSameInstant(defaultTimeZoneId);
       return Date.valueOf(targetZonedDateTime.toLocalDate());
     }
+    if (dateObj instanceof Timestamp) {
+      Timestamp timestamp = (Timestamp) dateObj;
+      if (cal == null) return new Date(timestamp.getTime());
+      long millis = timestamp.getTime();
+      long adjustedMillis = millis - cal.getTimeZone().getOffset(millis)
+          + TimeZone.getDefault().getOffset(millis);
```
Comment (QuChen88) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/cache/CachedResultSet.java]: To avoid doing `TimeZone.getDefault()` each time, put this on top of this class as a static field similar to `defaultTimeZoneId` on line 71.

File: wrapper/src/main/java/software/amazon/jdbc/plugin/cache/CachedResultSet.java [hunk 2]
```diff
@@ -302,6 +311,17 @@ private Time convertToTime(Object timeObj, Calendar cal) throws SQLException {
       OffsetTime localTime = ((OffsetTime)timeObj).withOffsetSameInstant(OffsetDateTime.now().getOffset());
       return Time.valueOf(localTime.toLocalTime());
     }
+    if (timeObj instanceof Timestamp) {
+      Timestamp timestamp = (Timestamp) timeObj;
+      if (cal == null) {
+        return new Time(timestamp.getTime());
+      } else {
```
Comment (QuChen88) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/cache/CachedResultSet.java]: nitpick: 

```
long millis = timestamp.getTime();
if (cal == null) return new Time(millis);
long adjustedMillis = ...
```

File: wrapper/src/main/java/software/amazon/jdbc/plugin/cache/CachedResultSet.java [hunk 3]
```diff
@@ -268,6 +269,14 @@ private Date convertToDate(Object dateObj, Calendar cal) throws SQLException {
       ZonedDateTime targetZonedDateTime = originalZonedDateTime.withZoneSameInstant(defaultTimeZoneId);
       return Date.valueOf(targetZonedDateTime.toLocalDate());
     }
+    if (dateObj instanceof Timestamp) {
+      Timestamp timestamp = (Timestamp) dateObj;
+      if (cal == null) return new Date(timestamp.getTime());
+      long millis = timestamp.getTime();
```
Comment (QuChen88) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/cache/CachedResultSet.java:276]: nitpick: Move this above the `if` above

File: wrapper/src/test/java/software/amazon/jdbc/plugin/cache/CachedResultSetTest.java [hunk 4]
```diff
@@ -13,8 +14,12 @@
 import org.junit.jupiter.api.Test;
 import org.mockito.Mock;
 import org.mockito.MockitoAnnotations;
+import org.testcontainers.shaded.org.bouncycastle.util.Times;
```
Comment (QuChen88) [inline wrapper/src/test/java/software/amazon/jdbc/plugin/cache/CachedResultSetTest.java]: What is this dependency?

File: wrapper/src/test/java/software/amazon/jdbc/plugin/cache/CachedResultSetTest.java [hunk 5]
```diff
@@ -368,10 +373,15 @@ void test_get_special_time() throws SQLException {
         4362000L,
         LocalTime.of(10, 20, 30),
         OffsetTime.of(12, 15, 30, 0, ZoneOffset.UTC),
+        new Timestamp(1755621000000L), // Date and time (GMT): Tuesday, August 19, 2025 4:30:00 PM
+        new Timestamp(1735713000000L), // Date and time (GMT): Wednesday, January 1, 2025 6:30:00 AM
+        new Timestamp(0L), // 1970-01-01 00:00:00 UTC (epoch)
+        new Timestamp(Long.MAX_VALUE), // Future date
```
Comment (QuChen88) [inline wrapper/src/test/java/software/amazon/jdbc/plugin/cache/CachedResultSetTest.java]: This is too far into the future.... Maybe make it happen like a year from now?

File: wrapper/src/test/java/software/amazon/jdbc/plugin/cache/CachedResultSetTest.java [hunk 6]
```diff
@@ -385,6 +395,33 @@ void test_get_special_time() throws SQLException {
     assertTrue(cachedRs.next());
     assertEquals(Time.valueOf("05:15:30"), cachedRs.getTime(1));
     assertEquals(Time.valueOf("05:15:30"), cachedRs.getTime(1, estCal));
+    // Time from Timestamp
+    assertTrue(cachedRs.next());
+    Timestamp timestampOne = new Timestamp(1755621000000L);
+    assertEquals(new Time(timestampOne.getTime()), cachedRs.getTime(1));
```
Comment (QuChen88) [inline wrapper/src/test/java/software/amazon/jdbc/plugin/cache/CachedResultSetTest.java]: `Time.valueOf()` is easier for someone to understand the expected time result you are testing against.

File: wrapper/src/test/java/software/amazon/jdbc/plugin/cache/CachedResultSetTest.java [hunk 7]
```diff
@@ -385,6 +395,33 @@ void test_get_special_time() throws SQLException {
...
+    assertEquals("09:30:00", cachedRs.getTime(1).toString());
+    assertEquals("06:30:00", cachedRs.getTime(1, estCal).toString());
```
Comment (QuChen88) [inline wrapper/src/test/java/software/amazon/jdbc/plugin/cache/CachedResultSetTest.java]: I would avoid `toString()` here. Instead just do `assertEquals(Time.valueOf("09:30:00"), cachedRs.getTime(1))`

File: wrapper/src/test/java/software/amazon/jdbc/plugin/cache/CachedResultSetTest.java [hunk 8]
```diff
@@ -385,6 +395,33 @@ void test_get_special_time() throws SQLException {
...
+    // Time from Timestamp Edge Case
+    assertTrue(cachedRs.next());
+    Timestamp timestampTwo = new Timestamp(1735713000000L);
+    long timestampMillis = timestampTwo.getTime();
+    long calOffset = estCal.getTimeZone().getOffset(timestampMillis);
+    long jvmDefaultOffset = TimeZone.getDefault().getOffset(timestampMillis);
+    long expectedTimeMillis = timestampMillis - calOffset + jvmDefaultOffset;
```
Comment (QuChen88) [inline wrapper/src/test/java/software/amazon/jdbc/plugin/cache/CachedResultSetTest.java]: No need to do this time zone conversion. I already set up the default JVM time zone to be PST (i.e. "America/Los_Angeles") in the `SetUp()` method, so you can safely test for `eastern time` being 3 hours ahead of the local JVM time.

File: wrapper/src/test/java/software/amazon/jdbc/plugin/cache/CachedResultSetTest.java [hunk 9]
```diff
@@ -424,9 +467,28 @@ void test_get_special_date() throws SQLException {
...
     assertEquals(new Date(-1000000000L), cachedRs.getDate(1));
     // Date from LocalDate
+
     assertTrue(cachedRs.next());
     assertEquals(Date.valueOf("2010-10-30"), cachedRs.getDate(1));
     assertEquals(Date.valueOf("2010-10-29"), cachedRs.getDate(1, estCal));
+    // Date from Timestamp
+    assertTrue(cachedRs.next());
+    assertEquals("2025-08-19", cachedRs.getDate(1).toString());
+    assertEquals("2025-08-19", cachedRs.getDate(1, estCal).toString());
```
Comment (QuChen88) [inline wrapper/src/test/java/software/amazon/jdbc/plugin/cache/CachedResultSetTest.java]: Use `Date.ValueOf()` to avoid `toString()`.

File: wrapper/src/test/java/software/amazon/jdbc/plugin/cache/CachedResultSetTest.java [hunk 10]
```diff
@@ -442,4 +504,283 @@ void test_get_special_date() throws SQLException {
...
+    // Test bytes values - both index and label versions
+    assertTrue(cachedRs.next());
+    assertArrayEquals(testBytes, cachedRs.getBytes(1));
+    assertFalse(cachedRs.wasNull());
+    assertArrayEquals(testBytes, cachedRs.getBytes("fieldByte"));
+    assertFalse(cachedRs.wasNull());
+
+    // Test non-byte array input
+    assertTrue(cachedRs.next());
+    assertArrayEquals(new byte[0], cachedRs.getBytes(1));
```
Comment (QuChen88) [inline wrapper/src/test/java/software/amazon/jdbc/plugin/cache/CachedResultSetTest.java]: Is this result / implementation correct when the content of the column is not a byte array type? :) 

The javadoc for `getBytes()`  https://docs.oracle.com/javase/8/docs/api/java/sql/ResultSet.html#getBytes-int-

File: wrapper/src/test/java/software/amazon/jdbc/plugin/cache/CachedResultSetTest.java [hunk 11]
```diff
@@ -442,4 +504,283 @@ void test_get_special_date() throws SQLException {
...
+    assertFalse(cachedRs.wasNull());
+
+    // Test number input
+    assertTrue(cachedRs.next());
+    assertArrayEquals(new byte[0], cachedRs.getBytes(1));
```
Comment (QuChen88) [inline wrapper/src/test/java/software/amazon/jdbc/plugin/cache/CachedResultSetTest.java]: Same here.

File: wrapper/src/test/java/software/amazon/jdbc/plugin/cache/CachedResultSetTest.java [hunk 12]
```diff
@@ -442,4 +504,283 @@ void test_get_special_date() throws SQLException {
...
+    assertTrue(cachedRs.next());
+    assertFalse(cachedRs.getBoolean(1));
+    assertFalse(cachedRs.wasNull());
+
+    // Test number conversions: 0 = true, non-zero = false
+    assertTrue(cachedRs.next());
+    assertTrue(cachedRs.getBoolean(1)); // 0 → true
+
+    assertTrue(cachedRs.next());
+    assertFalse(cachedRs.getBoolean(1)); // 1 → false
```
Comment (QuChen88) [inline wrapper/src/test/java/software/amazon/jdbc/plugin/cache/CachedResultSetTest.java]: `0` means false and `1` means true? 

https://docs.oracle.com/javase/8/docs/api/java/sql/ResultSet.html#getBoolean-int-

File: wrapper/src/test/java/software/amazon/jdbc/plugin/cache/CachedResultSetTest.java [hunk 13]
```diff
@@ -442,4 +504,283 @@ void test_get_special_date() throws SQLException {
...
+    when(mockResultSetMetadata.getColumnCount()).thenReturn(1);
+    mockGetMetadataFields(1, 2);
+
+    // Test data: string, integer, boolean, null
+    when(mockResultSet.getObject(1)).thenReturn("test", 123, true, null);
+    when(mockResultSet.next()).thenReturn(true, true, true, true, false);
+
+    CachedResultSet cachedRs = new CachedResultSet(mockResultSet);
+
+    // Test valid type conversions
```
Comment (QuChen88) [inline wrapper/src/test/java/software/amazon/jdbc/plugin/cache/CachedResultSetTest.java:730]: I would test for one of the data types that are not supported by the `getXXX()` APIs. i.e. `java.util.HashSet` or something...

File: wrapper/src/test/java/software/amazon/jdbc/plugin/cache/CachedResultSetTest.java (hunk 7 above)
Comment (Frank-Gu-81) [reply wrapper/src/test/java/software/amazon/jdbc/plugin/cache/CachedResultSetTest.java]: This is something that I was thinking about as well. From my understanding, the previous tests with `Time.valueOf()` works becuase we are returning a `Time.valueOf()` object for `LocalTime` and `OffsetTime`. However, for `Timestamp` object, I'm returning a new `Timestamp` object that would cause assertion to fail:

Expected :09:30:00
Actual :09:30:00
org.opentest4j.AssertionFailedError: expected: java.sql.Time@75f9a35a<09:30:00> but was: java.sql.Time@7acbd3d0<09:30:00>

Which is due to comparing different object instances with different memory addresses. Do you know if there's a better way for me to test without running into this issue?

Comment (QuChen88) [reply wrapper/src/test/java/software/amazon/jdbc/plugin/cache/CachedResultSetTest.java]: I would test the `milliseconds` value of `Time` is correct then. i.e. `assertEquals(X, cachedRs.getTime(1).getTime())`.

File: wrapper/src/main/java/software/amazon/jdbc/plugin/cache/CachedResultSet.java (hunk 3 above)
Comment (Frank-Gu-81) [reply wrapper/src/main/java/software/amazon/jdbc/plugin/cache/CachedResultSet.java:275]: Just to double-check, is this the structure you want? 

`if (cal == null && dateObj instanceof Timestamp) {
      Timestamp timestamp = (Timestamp) dateObj;
      return new Date(timestamp.getTime());
    }
    if (dateObj instanceof Timestamp) {
      Timestamp timestamp = (Timestamp) dateObj;
      long millis = timestamp.getTime();
      long adjustedMillis = millis - cal.getTimeZone().getOffset(millis)
          + defaultTimeZone.getOffset(millis);
      return new Date(adjustedMillis);
    }`

File: wrapper/src/test/java/software/amazon/jdbc/plugin/cache/CachedResultSetTest.java (hunk 10 above)
Comment (Frank-Gu-81) [reply wrapper/src/test/java/software/amazon/jdbc/plugin/cache/CachedResultSetTest.java]: Fixed it. It looks like the implementation is incorrect.

File: wrapper/src/test/java/software/amazon/jdbc/plugin/cache/CachedResultSetTest.java (hunk 12 above)
Comment (Frank-Gu-81) [reply wrapper/src/test/java/software/amazon/jdbc/plugin/cache/CachedResultSetTest.java]: Seems to be an implementation issue as well. Fixed it.

File: wrapper/src/test/java/software/amazon/jdbc/plugin/cache/CachedResultSetTest.java (hunk 7 above)
Comment (Frank-Gu-81) [reply wrapper/src/test/java/software/amazon/jdbc/plugin/cache/CachedResultSetTest.java]: Got it. Aside from the milliseconds level, I also added another assertion to check the wall clock time using `LocalTime`. 

```java
assertEquals(futureTimestamp.getTime(), cachedRs.getTime(1).getTime());
    assertEquals(LocalTime.of(9, 30, 0), cachedRs.getTime(1).toLocalTime());
```

File: wrapper/src/main/java/software/amazon/jdbc/plugin/cache/CachedResultSet.java (hunk 3 above)
Comment (Frank-Gu-81) [reply wrapper/src/main/java/software/amazon/jdbc/plugin/cache/CachedResultSet.java:276]: Just to double-check, is this the structure you are looking for?

```java
if (timeObj instanceof OffsetTime) {
  OffsetTime localTime = ((OffsetTime)timeObj).withOffsetSameInstant(OffsetDateTime.now().getOffset());
  return Time.valueOf(localTime.toLocalTime());
}
if (timeObj instanceof Timestamp) {
  Timestamp timestamp = (Timestamp) timeObj;
  if (cal == null) return new Time(timestamp.getTime());
  long millis = timestamp.getTime();
  long adjustedMillis = millis - cal.getTimeZone().getOffset(millis)
      + TimeZone.getDefault().getOffset(millis);
  return new Time(adjustedMillis);
}
```

File: wrapper/src/test/java/software/amazon/jdbc/plugin/cache/CachedResultSetTest.java [hunk 14]
```diff
@@ -368,10 +370,15 @@ void test_get_special_time() throws SQLException {
         4362000L,
         LocalTime.of(10, 20, 30),
         OffsetTime.of(12, 15, 30, 0, ZoneOffset.UTC),
+        new Timestamp(1755621000000L), // Date and time (GMT): Tuesday, August 19, 2025 4:30:00 PM
+        new Timestamp(1735713000000L), // Date and time (GMT): Wednesday, January 1, 2025 6:30:00 AM
+        new Timestamp(0L), // 1970-01-01 00:00:00 UTC (epoch)
+        new Timestamp(1787416200000L), // Future Date: Date and time (GMT): Saturday, August 22, 2026 4:30:00 PM
```
Comment (QuChen88) [inline wrapper/src/test/java/software/amazon/jdbc/plugin/cache/CachedResultSetTest.java]: I meant you make this one year from `now`. 

If you hard-code it to 2026 August, this test will no longer be testing a future date once we get to Aug 2026 :)

File: wrapper/src/main/java/software/amazon/jdbc/plugin/cache/CachedResultSet.java [hunk 15]
```diff
@@ -302,6 +313,14 @@ private Time convertToTime(Object timeObj, Calendar cal) throws SQLException {
       OffsetTime localTime = ((OffsetTime)timeObj).withOffsetSameInstant(OffsetDateTime.now().getOffset());
       return Time.valueOf(localTime.toLocalTime());
     }
+    if (timeObj instanceof Timestamp) {
+      Timestamp timestamp = (Timestamp) timeObj;
+      long millis = timestamp.getTime();
+      if (cal == null) return new Time(millis);
+      long adjustedMillis = millis - cal.getTimeZone().getOffset(millis)
+          + TimeZone.getDefault().getOffset(millis);
```
Comment (QuChen88) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/cache/CachedResultSet.java]: `defaultTimeZone.getOffset()`
